File: src/tools/insights/getAll.ts
import type { z } from "zod";
import type { Context, Tool } from "../types";
import { getProjectBaseUrl } from "../../lib/utils/api";
import { InsightGetAllSchema } from "../../schema/tool-inputs";

const schema = InsightGetAllSchema;

type Params = z.infer<typeof schema>;

export const getAllHandler = async (context: Context, params: Params) => {
	const { data } = params;
	const projectId = await context.getProjectId();
	const insightsResult = await context.api.insights({ projectId }).list({ params: data });

	if (!insightsResult.success) {
		throw new Error(`Failed to get insights: ${insightsResult.error.message}`);
	}

	const insightsWithUrls = insightsResult.data.map((insight) => ({
		...insight,
		url: `${getProjectBaseUrl(projectId)}/insights/${insight.short_id}`,
	}));

	return { content: [{ type: "text", text: JSON.stringify(insightsWithUrls) }] };
};

const tool = (): Tool<typeof schema> => ({
	name: "insights-get-all",
	description: `
        - Get all insights in the project with optional filtering.
        - Can filter by saved status, favorited status, or search term.
    `,
	schema,
	handler: getAllHandler,
});

export default tool;

File: python/tools/insights/get_all.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.insights import ListInsights
from schema.tool_inputs import InsightGetAllSchema
from tools.types import Context, Tool, ToolResult


async def get_all_insights_handler(context: Context, params: InsightGetAllSchema) -> ToolResult:
    project_id = await context.get_project_id()

    data = params.data

    if data:
        list_params = ListInsights(
            limit=int(data.limit) if data.limit else None,
            offset=int(data.offset) if data.offset else None,
            saved=data.saved,
            favorited=data.favorited,
            search=data.search,
        )
    else:
        list_params = ListInsights()

    insights_result = await context.api.insights(project_id).list(list_params)

    if is_error(insights_result):
        raise Exception(f"Failed to get insights: {insights_result.error}")

    assert is_success(insights_result)

    insights_with_urls = []
    for insight in insights_result.data:
        insight_dict = insight.model_dump()
        insight_dict["url"] = f"{get_project_base_url(project_id, context.api.base_url)}/insights/{insight.short_id}"
        insights_with_urls.append(insight_dict)

    return ToolResult(content=json.dumps(insights_with_urls))


def get_all_insights_tool() -> Tool[InsightGetAllSchema]:
    return Tool(
        name="insights-get-all",
        description="Use this tool to get all insights for the project. Supports filtering by saved, favorited, and search query.",
        schema=InsightGetAllSchema,
        handler=get_all_insights_handler,
    )
